### ISSUE TITLE
About feature extraction after reduced dimension of components

### ISSUE THREAD

**Issue by csitaula** [jlmelville/uwot#8]
I have training and testing datasets and I reduced their dimension separately (for training and testing data) and accuracy dropped in SVM. What to do? Is it necessary to keep all training data in a single matrix and reduce their dimension for reduced feature extraction? Thanks

**Comment by jlmelville** [comment]
You definitely don't want to do dimensionality reduction on the training and test sets separately (you wouldn't want to do this with a traditional unsupervised method like PCA either). Consider using the metric learning features of UMAP:

```R
umap_model <- umap(training_data, ret_model = TRUE)
umap_test <- umap_transform(test_data, umap_model)
```
You must use `ret_model = TRUE` which returns more data needed for the `umap_transform` step. The training data embedded coordinates are in `umap_model$embedding` and the test data coordinates are `umap_test`. There's an example using Fashion MNIST at https://jlmelville.github.io/uwot/metric-learning.html#metric_learning.

Of course, there's no guarantee that your SVM results won't still show reduced accuracy, but good luck!

**Comment by csitaula** [comment]
Is it necessary to normalise once I extract features (components)from umap
before feeding into svm ? thanks

**Comment by jlmelville** [comment]
I don't know. You are one of the first people to try this as far as I know. I have no idea even how much the metric learning features of the [official Python implementation](https://umap-learn.readthedocs.io/en/latest/supervised.html#training-with-labels-and-embedding-unlabelled-test-data-metric-learning-with-umap) have been used, but it's all fairly new.

For what it's worth (which isn't very much), I would simply use whatever normalization procedure you carry out on the raw data when applying an SVM to it. All other things being equal, my preference would be to normalize the data before applying UMAP, then avoid further normalization of the UMAP output when it was used as input to SVM.

**Comment by csitaula** [comment]
thanks you so much! let me try this!

On 10/6/18, James Melville <notifications@github.com> wrote:
> I don't know. You are one of the first people to try this as far as I know.
> I have no idea even how much the metric learning features of the [official
> Python
> implementation](https://umap-learn.readthedocs.io/en/latest/supervised.html#training-with-labels-and-embedding-unlabelled-test-data-metric-learning-with-umap)
> have been used, but it's all fairly new.
>
> For what it's worth (which isn't very much), I would simply use whatever
> normalization procedure you carry out on the raw data when applying an SVM
> to it. All other things being equal, my preference would be to normalize the
> data before applying UMAP, then avoid further normalization of the UMAP
> output when it was used as input to SVM.
>
>
> --
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/jlmelville/uwot/issues/8#issuecomment-427533032


-- 
******************************************
With Love,
Chiranjibi Sitaula
Postgraduate Research Fellow
Department of IT( Faculty of Science and Engineering)
Deakin University,Melbourne, Australia
ORCID:orcid.org/0000-0002-4564-2985
CYA E-RYT 550 Certified Yoga Instructor from Nepal.

**Comment by csitaula** [comment]
I found no such reduction in accuracy even after reduction of dimension
using SVM. I think this tool can be be used to aggregate multiple features
of high dimensional data and reduce to extract new types of features.
Thanks!
******************************************
With Love,
Chiranjibi Sitaula
Postgraduate Research Fellow
Department of IT( Faculty of Science and Engineering)
Deakin University,Melbourne, Australia
ORCID:orcid.org/0000-0002-4564-2985
CYA E-RYT 550 Certified Yoga Instructor from Nepal.

**Comment by jlmelville** [comment]
Glad you have found it useful!
